docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,212 @@
+4. `GCM`:
+    - During reconcile of the `Bastion` resource:
+      - according to `spec.shootRef`, sets the `status.seedName`
+      - according to `spec.shootRef`, sets the `status.providerType`
+5. Gardener extension provider <infra> / Bastion Controller on Seed:
+    - With own `Bastion` Custom Resource Definition in the seed under the api group `extensions.gardener.cloud/v1alpha1`
+    - Watches `Bastion` custom resources that are created by the `gardenlet` in the seed
+    - Controller reads `cloudprovider` credentials from seed-shoot namespace
+    - Deploy infrastructure resources
+        - Bastion VM. User data similar to https://github.com/gardener/gardenctl/blob/1e3e5fa1d5603e2161f45046ba7c6b5b4107369e/pkg/cmd/ssh.go#L160-L171. Writes `spec.sshPublicKey` into `authorized_keys` file.
+        - attaches public IP, creates security group, firewall rules, etc.
+    - Updates status of `Bastion` resource:
+        - With bastion IP under `status.ingress.ip` or hostname under `status.ingress.hostname`
+        - Updates the `status.lastOperation` with the status of the last reconcile operation
+6. `gardenlet`
+    - Syncs back the `status.ingress` and `status.conditions` of the `Bastion` resource in the seed to the garden cluster in case it changed
+7. `gardenctl`
+    - initiates `ssh` session once `status.conditions['BastionReady']` is true of the `Bastion` resource in the garden cluster
+        - locates private `ssh` key matching `spec["sshPublicKey"]` which was configured beforehand by the user
+        - reads bastion IP (`status.ingress.ip`) or hostname (`status.ingress.hostname`)
+        - reads the private key from the `ssh` key pair for the shoot node
+        - opens `ssh` connection to the bastion and from there to the respective shoot node
+    - runs heartbeat in parallel as long as the `ssh` session is open by annotating the `Bastion` resource with `operations.gardener.cloud/operation: keepalive`
+8. `GCM`:
+    - Once `status.expirationTimestamp` is reached, the `Bastion` will be marked for deletion
+9. `gardenlet`:
+    - Once the `Bastion` resource in the garden cluster is marked for deletion, it marks the `Bastion` resource in the seed for deletion
+10. Gardener extension provider <infra> / Bastion Controller on Seed:
+    - all created resources will be cleaned up
+    - On succes, removes finalizer on `Bastion` resource in seed
+11. `gardenlet`:
+    - removes finalizer on `Bastion` resource in garden cluster
+
+### Resource Example
+
+`Bastion` resource in the garden cluster
+```yaml
+apiVersion: operations.gardener.cloud/v1alpha1
+kind: Bastion
+metadata:
+  generateName: cli-
+  name: cli-abcdef
+  namespace: garden-myproject
+  annotations:
+    operations.gardener.cloud/created-by: foo # set by the mutating webhook
+    # operations.gardener.cloud/operation: keepalive # this annotation is removed by the GAPI and the last-heartbeat timestamp and the status.expirationTimestamp will be updated accordingly
+spec:
```

Comment:
Have you decided which portion of the spec should be immutable, the whole `spec` vs only certain fields? Can you add it as a comment?

Reply:
Good point, somehow some of them got lost so I added them again. See diff https://github.com/gardener/gardener/pull/3802/commits/268d9dd9556b68b0458f160866d90b1e6566a6ed

@vlerenc I added the validation for `spec.ingress` again:

```
### SSH Flow
[...]
2. GAPI Admission Plugin for the `Bastion` resource in the garden cluster
[...]
    - validates that only the creator of the bastion (see `gardener.cloud/created-by` annotation) can update `spec.ingress`
```

I can't recall why I have removed a similar validation sentence during one of the pr feedback commits but my gut feeling is that we should validate this, even though the `spec.sshPublicKey` is immutable, there should be no reason why any other user should be allowed to change this field, except for the owner / creator.

Reply:
@petersutter Sorry, maybe I deleted this accidentally when I (too quickly) tried to rid us of the intermediate keypair and it's encryption with the GPG key.